pkg/machinehealthcheck/machinehealthcheck.go

```diff
@@ -68,57 +49,105 @@ var _ = Describe("[Feature:MachineHealthCheck] MachineHealthCheck", func() {
 	})
 
 	AfterEach(func() {
+		By("Deleting the MachineHealthCheck resource")
+		err := client.Delete(context.Background(), machinehealthcheck)
+		Expect(err).ToNot(HaveOccurred())
+
 		By("Deleting the new MachineSet")
-		err := client.Delete(context.Background(), machineSet)
+		err = client.Delete(context.Background(), machineSet)
 		Expect(err).ToNot(HaveOccurred())
 
 		framework.WaitForMachineSetDelete(client, machineSet)
 	})
 
 	It("should remediate unhealthy nodes", func() {
-		By("Setting conditions on a Node")
-
 		selector := machineSet.Spec.Selector
 		machines, err := framework.GetMachines(client, &selector)
 		Expect(err).ToNot(HaveOccurred())
 		Expect(machines).ToNot(BeEmpty())
 
-		machine := machines[0]
-		node, err := framework.GetNodeForMachine(client, machine)
-		Expect(err).ToNot(HaveOccurred())
-
-		err = framework.AddNodeCondition(client, node, nodeCondition)
-		Expect(err).ToNot(HaveOccurred())
+		By("Setting unhealthy conditions on machine nodes, but not exceding maxUnhealthy threshold")
+		unhealthyMachines := machines[:maxUnhealthy-1]
+		for _, machine := range unhealthyMachines {
+			node, err := framework.GetNodeForMachine(client, machine)
+			Expect(err).ToNot(HaveOccurred())
+			err = framework.AddNodeCondition(client, node, nodeCondition)
+			Expect(err).ToNot(HaveOccurred())
+		}
 
 		By("Creating a MachineHealthCheck resource")
 		mhcParams := framework.MachineHealthCheckParams{
-			Name:   machineSetParams.Name,
-			Labels: machineSetParams.Labels,
+			Name:   machineSet.Name,
+			Labels: machineSet.Labels,
 			Conditions: []mapiv1beta1.UnhealthyCondition{
 				{
 					Type:    E2EConditionType,
 					Status:  corev1.ConditionTrue,
 					Timeout: "1s",
 				},
 			},
+			MaxUnhealthy: &maxUnhealthy,
 		}
 
-		mhc, err := framework.CreateMHC(client, mhcParams)
+		machinehealthcheck, err = framework.CreateMHC(client, mhcParams)
 		Expect(err).ToNot(HaveOccurred())
-		Expect(mhc).ToNot(BeNil())
+		Expect(machinehealthcheck).ToNot(BeNil())
 
-		By("Verifying the matching Machine is deleted")
-		framework.WaitForMachineDelete(client, machine)
+		By("Waiting for each unhealthy machine to be deleted")
+		framework.WaitForMachinesDeleted(client, unhealthyMachines...)
+
+		By("Waiting for MachineDeleted event from MachineHealthCheck for each unhealthy machine")
+		for _, machine := range unhealthyMachines {
+			err = framework.WaitForEvent(client, "Machine", machine.Name, "MachineDeleted")
+			Expect(err).ToNot(HaveOccurred())
```

Comment:
Could inline this? That way we don't shadow the err variable
```suggestion
			Expect(framework.WaitForEvent(client, "Machine", machine.Name, "MachineDeleted")).To(Succeed())
```